File: 46-permutations/permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        
        List<List<Integer>> ans = new ArrayList<>();

        boolean[] map = new boolean[nums.length];
        Arrays.fill(map,false);

        helper(nums, map, new ArrayList<Integer>(), ans);
        return ans;

    }

    void helper(int[] nums, boolean[] map, List<Integer> ds, List<List<Integer>> ans){

        if (ds.size() == nums.length){
            ans.add(new ArrayList<>(ds));
            return;
        }

        for (int i = 0;i<nums.length;i++){
            if (!map[i]){
            ds.add(nums[i]);
            map[i] = true;

             helper(nums, map, ds, ans);
            ds.remove(ds.size()-1);
            map[i] = false;
            }

           
            
            
        }
    }
}

File: 46-permutations/permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        
        cur = []
        ans = []
        vis = [-1]*len(nums)
        self.helper(0, nums, cur, ans, vis)

        return ans 

    def helper(self, ind, nums, cur, ans, vis):

        if len(cur) == len(nums):
            ans.append(cur.copy())
            return

        for i in range(len(nums)):
            num = nums[i]
            if vis[i] == -1:
                cur.append(num)
                vis[i] = 1
                self.helper(ind, nums, cur, ans, vis)
                vis[i] = -1
                cur.pop()
